File: apps/api/app/models/dictation.py
"""ORM model: dictation — DATA_MODEL.md §Core Tables.
Pipeline: TECHNICAL_DESIGN.md §4.
Override log (overridden[]) is an eval asset — never discard.
"""
import uuid
from sqlalchemy import String, ForeignKey, Text, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class Dictation(Base):
    __tablename__ = "dictation"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    visit_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("visit.id", ondelete="CASCADE"),
        nullable=False, index=True
    )
    # Cloud storage reference to original audio file
    audio_ref: Mapped[str | None] = mapped_column(String(512), nullable=True)
    language: Mapped[str] = mapped_column(String(10), nullable=False, default="en")
    transcript: Mapped[str] = mapped_column(Text, nullable=False)

    # JSONB: list of Extraction objects
    # Each: {id, source_phrase, term_code, term_en, target_field, suggested_value, confidence}
    extractions: Mapped[list] = mapped_column(JSON, nullable=False, default=list)

    # IDs of extractions accepted without change
    accepted: Mapped[list] = mapped_column(JSON, nullable=False, default=list)

    # Override log: list of {extraction_id, action, original_value, final_value, reviewed_at}
    # action: accept|edit|reject
    # THIS IS AN EVAL ASSET — never delete/truncate per DATA_MODEL.md
    overridden: Mapped[list] = mapped_column(JSON, nullable=False, default=list)

    created_at: Mapped[str] = mapped_column(String(30), nullable=False, server_default="now()")

    visit: Mapped["Visit"] = relationship("Visit", back_populates="dictations")  # type: ignore[name-defined]  # noqa: F821

File: packages/schema/src/dictation.ts
// Dictation — DATA_MODEL.md §Core Tables: `dictation`
// Pipeline: TECHNICAL_DESIGN.md §4
// Override log is an eval asset — never discard
import type { ReviewAction } from "./enums";

export interface Extraction {
  id: string;
  /** Span of text that matched */
  source_phrase: string;
  /** Matched terminology code */
  term_code: string;
  term_en: string;
  /** Target field path, e.g. "clinical_exam.pulse.quality" */
  target_field: string;
  suggested_value: string;
  confidence: number;
}

export interface Dictation {
  id: string;
  visit_id: string;
  /** Cloud storage reference to original audio */
  audio_ref?: string;
  language: string;
  transcript: string;
  extractions: Extraction[];
  /** IDs of extractions accepted without change */
  accepted: string[];
  /** Override log: extraction_id → { action, original, final_value } */
  overridden: OverrideLogEntry[];
  created_at: string;
}

export interface OverrideLogEntry {
  extraction_id: string;
  action: ReviewAction;
  original_value: string;
  final_value?: string;
  reviewed_at: string;
}

/** POST /dictation/{id}/review payload */
export interface ReviewPayload {
  extraction_id: string;
  action: ReviewAction;
  /** Required when action === 'edit' */
  final_value?: string;
}

/** Job response after POST /visits/{id}/dictation */
export interface DictationJob {
  job_id: string;
  status: "queued" | "processing" | "done" | "error";
  dictation?: Dictation;
  error?: string;
}
